server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/VacationMailet.java

```diff
@@ -0,0 +1,116 @@
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.mailet;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import javax.inject.Inject;
+import javax.mail.MessagingException;
+
+import org.apache.james.jmap.api.vacation.AccountId;
+import org.apache.james.jmap.api.vacation.Vacation;
+import org.apache.james.jmap.api.vacation.VacationRepository;
+import org.apache.james.jmap.utils.ZonedDateTimeProvider;
+import org.apache.mailet.Mail;
+import org.apache.mailet.MailAddress;
+import org.apache.mailet.base.GenericMailet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class VacationMailet extends GenericMailet {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(VacationMailet.class);
+
+    public static class ContextualizedValue<T> {
+        private final MailAddress recipient;
+        private final T value;
+
+        public ContextualizedValue(MailAddress recipient, T value) {
+            this.recipient = recipient;
+            this.value = value;
+        }
+
+        public MailAddress getRecipient() {
+            return recipient;
+        }
+
+        public T getValue() {
+            return value;
+        }
+    }
+
+    private final VacationRepository vacationRepository;
+    private final ZonedDateTimeProvider zonedDateTimeProvider;
+
+    @Inject
+    public VacationMailet(VacationRepository vacationRepository, ZonedDateTimeProvider zonedDateTimeProvider) {
+        this.vacationRepository = vacationRepository;
+        this.zonedDateTimeProvider = zonedDateTimeProvider;
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        mail.getRecipients()
+            .stream()
+            .map(mailAddress -> new ContextualizedValue<>(mailAddress,
+                AccountId.create(mailAddress.getLocalPart() + "@" + mailAddress.getDomain())))
+            .map(accountIdWithContext -> new ContextualizedValue<>(accountIdWithContext.getRecipient(),
+                vacationRepository.retrieveVacation(accountIdWithContext.getValue())))
+            .map(vacationWithContext -> manageVacation(vacationWithContext, mail))
+            .forEach(CompletableFuture::join);
+    }
+
+    public CompletableFuture<Vacation> manageVacation(ContextualizedValue<CompletableFuture<Vacation>> contextualizedVacation, Mail processedMail) {
+        return contextualizedVacation.getValue().whenCompleteAsync((vacation, exception) -> {
+            if (vacation != null) {
+                if (shouldSendNotification(vacation, processedMail, contextualizedVacation.getRecipient())) {
+                    sendNotification(contextualizedVacation.getRecipient(), processedMail, vacation);
+                }
+            } else {
+                LOGGER.warn("Can not retrieve vacation", exception);
```

Comment:
<irony>I love the way to handle futures stage errors </irony>

We do not get vacations when we encounter a Cassandra error, avoinding us to return the stored value nor the default value. Instead it throws a subclass of runtimeException. I thought it would be nice to handle such exceptions, even if the syntax is uggly.